File: server/src/types/growth.ts
/**
 * Growth Module — Domain Types & Validation Schemas
 *
 * This file defines the core data contracts for the baby growth tracking module.
 * It serves as the single source of truth for shape and validation rules consumed
 * by controllers, services, and the database layer.
 *
 * Architecture role:
 * - **Validation boundary**: Zod schemas enforce structural and semantic rules at
 *   the API entry point before data reaches business logic or persistence.
 * - **Type derivation**: All "Create" / "Update" TypeScript types are inferred from
 *   their Zod schemas via `z.infer`, guaranteeing that runtime validation and
 *   compile-time type-checking stay in lockstep.
 * - **Read-model interfaces**: `GrowthRecord`, `Child`, and `ChildSummary` describe
 *   the shapes returned by the database / service layer — they carry enriched data
 *   (generated IDs, timestamps) that the input schemas do not include.
 *
 * Conventions:
 * - Input DTOs (Create / Update variants) are **Zod-inferred** — they never drift from the
 *   validation rules.
 * - Output / persisted models are **explicit TypeScript interfaces** — they contain
 *   server-generated fields (id, childId, createdAt, updatedAt) that clients
 *   must not supply.
 * - All date strings use the `YYYY-MM-DD` format (ISO 8601 date-only).
 */

import { z } from 'zod';

/* ====================================================================
   Zod Validation Schemas
   These define the runtime validation contracts for incoming API payloads.
   Every constraint (min, max, regex) mirrors domain invariants.
   ==================================================================== */

/**
 * Feeding — describes a single feeding event recorded inside a growth record.
 *
 * Represents what and how much the baby consumed during a feeding session.
 * The `amount` field is optional to accommodate "on-demand" breastfeeding
 * where a precise measurement is not always available.
 */
export const FeedingSchema = z.object({
  /** Feeding method. `breast` = breastmilk, `formula` = prepared formula, `solid` = solid food, `mixed` = combination */
  type: z.enum(['breast', 'formula', 'solid', 'mixed']),

  /**
   * Quantity consumed. Omitted when the amount could not be measured
   * (e.g. on-demand nursing).
   */
  amount: z.number().positive().optional(),

  /**
   * Unit of measurement for the amount.
   * - `ml` / `oz` — liquids (breastmilk, formula)
   * - `g` — solids and semi-solids
   * - `次` — counting-based measure ("times"), used when volume/weight is impractical
   */
  unit: z.enum(['ml', 'oz', 'g', '次']).optional(),

  /**
   * Free-text notes about this feeding (e.g. "refused solids", "allergic reaction to egg").
   * Capped at 200 characters to prevent abuse of free-text fields.
   */
  notes: z.string().max(200).optional(),
});

/**
 * CreateGrowthRecord — input DTO for recording a new growth measurement.
 *
 * Every field except `date` is optional: a caregiver may record only the data points
 * they have at hand (e.g. weight-only visit to the paediatrician).
 *
 * Validation bounds reflect WHO child growth standards for children 0–5 years:
 * - Height: 30–180 cm (covers preemie to tall 5-year-old)
 * - Weight: 1–100 kg (covers low-birthweight to obese 5-year-old)
 * - Head circumference: 20–80 cm (microcephaly → macrocephaly range)
 * - Sleep: 0–24 hours (full-day cap; sum of naps + night sleep)
 * - Diapers: 0–50 per day (high-end accommodates newborn frequency)
 */
export const CreateGrowthRecordSchema = z.object({
  /** Measurement date in YYYY-MM-DD format (ISO 8601 date-only). */
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, '日期格式: YYYY-MM-DD'),

  /** Height in cm. WHO range: 30 cm (extreme preemie) to 180 cm (tall 5-year-old). */
  height: z.number().min(30).max(180).optional(),

  /** Weight in kg. Acceptable range: 1 kg (low birthweight) to 100 kg (extreme upper bound). */
  weight: z.number().min(1).max(100).optional(),

  /** Head circumference in cm. Clinical range: 20 cm (microcephaly) to 80 cm (macrocephaly). */
  headCircumference: z.number().min(20).max(80).optional(),

  /** Total sleep duration in hours (aggregate of naps + night sleep). Capped at 24h. */
  sleepDuration: z.number().min(0).max(24).optional(),

  /** Feeding details for this record. See {@link FeedingSchema}. */
  feeding: FeedingSchema.optional(),

  /** Number of diaper changes. 0–50 range accommodates newborn frequency (up to ~12/day). */
  diapers: z.number().int().min(0).max(50).optional(),

  /** Free-text caregiver notes. Capped at 500 characters. */
  notes: z.string().max(500).optional(),
});

/**
 * UpdateGrowthRecord — partial-update DTO for modifying an existing growth record.
 *
 * Derived from `CreateGrowthRecordSchema.partial()` — every field, including `date`,
 * becomes optional so callers can send only the fields they want to change (PATCH semantics).
 */
export const UpdateGrowthRecordSchema = CreateGrowthRecordSchema.partial();

/**
 * CreateChild — input DTO for registering a new baby/child profile.
 *
 * The `birthDate` is required at creation time because it anchors all age-based
 * percentile calculations (WHO growth charts compare measurements to exact age in months).
 */
export const CreateChildSchema = z.object({
  /** Display name for the child (1–50 characters). */
  name: z.string().min(1).max(50),

  /** Date of birth in YYYY-MM-DD format. Drives age-at-measurement calculations for percentile lookups. */
  birthDate: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, '日期格式: YYYY-MM-DD'),

  /** Biological sex — used to select the correct WHO reference population (male/female growth charts differ). */
  gender: z.enum(['male', 'female']),
});

/**
 * UpdateChild — partial-update DTO for modifying a child profile.
 *
 * Derived from `CreateChildSchema.partial()`. All fields optional (PATCH semantics).
 */
export const UpdateChildSchema = CreateChildSchema.partial();

/* ====================================================================
   TypeScript Types (inferred from Zod schemas)
   These are the compile-time companions to the runtime validation above.
   ==================================================================== */

/** Shape of a single feeding event. Inferred from {@link FeedingSchema}. */
export type Feeding = z.infer<typeof FeedingSchema>;

/** Shape of the request body for `POST /growth-records`. Inferred from {@link CreateGrowthRecordSchema}. */
export type CreateGrowthRecord = z.infer<typeof CreateGrowthRecordSchema>;

/** Shape of the request body for `PATCH /growth-records/:id`. Every field is optional. Inferred from {@link UpdateGrowthRecordSchema}. */
export type UpdateGrowthRecord = z.infer<typeof UpdateGrowthRecordSchema>;

/** Shape of the request body for `POST /children`. Inferred from {@link CreateChildSchema}. */
export type CreateChild = z.infer<typeof CreateChildSchema>;

/** Shape of the request body for `PATCH /children/:childId`. Every field is optional. Inferred from {@link UpdateChildSchema}. */
export type UpdateChild = z.infer<typeof UpdateChildSchema>;

/* ====================================================================
   Read-Model Interfaces (output / persisted shapes)
   These are the shapes returned by the service layer and stored in the database.
   They include server-generated fields that input DTOs intentionally exclude.
   ==================================================================== */

/**
 * GrowthRecord — the persisted representation of a single growth measurement.
 *
 * Returned by:
 * - `GET /children/:childId/records`
 * - `GET /children/:childId/records/:id`
 * - As elements inside `Child.records[]`
 *
 * Differs from `CreateGrowthRecord` by adding server-assigned `id` and `createdAt`.
 */
export interface GrowthRecord {
  /** Unique identifier assigned by the persistence layer on creation. */
  id: string;

  /** Measurement date in YYYY-MM-DD format. */
  date: string;

  /** Height in cm. Absent when not measured at this visit. */
  height?: number;

  /** Weight in kg. Absent when not measured at this visit. */
  weight?: number;

  /** Head circumference in cm. Absent when not measured at this visit. */
  headCircumference?: number;

  /** Total sleep in hours. Absent when not tracked for this record. */
  sleepDuration?: number;

  /** Feeding details. Absent when feeding was not recorded. */
  feeding?: Feeding;

  /** Number of diaper changes. Absent when not tracked. */
  diapers?: number;

  /** Free-text caregiver notes. Absent when no notes were provided. */
  notes?: string;

  /** ISO 8601 timestamp of when this record was first persisted. */
  createdAt: string;
}

/**
 * Child — the full representation of a child profile including all growth records.
 *
 * Returned by:
 * - `GET /children/:childId`
 * - `POST /children` (the created child, with empty `records` initially)
 *
 * This is the "fat" read model — `records` is populated eagerly.
 * For list endpoints that must avoid loading full record histories,
 * use {@link ChildSummary} instead.
 */
export interface Child {
  /** Unique identifier assigned by the persistence layer on creation. */
  childId: string;

  /** Display name for the child. */
  name: string;

  /** Date of birth in YYYY-MM-DD format. */
  birthDate: string;

  /** Biological sex — drives WHO percentile chart selection. */
  gender: 'male' | 'female';

  /** All growth records associated with this child, ordered by date descending (most recent first). */
  records: GrowthRecord[];

  /** ISO 8601 timestamp of when this child profile was first persisted. */
  createdAt: string;

  /** ISO 8601 timestamp of the last update to this child profile. */
  updatedAt: string;
}

/**
 * ChildSummary — a lightweight projection of a child profile for list views.
 *
 * Returned by:
 * - `GET /children` (list all children)
 *
 * Omits the full `records` array to keep list responses small. Includes
 * `recordCount` and `lastRecordDate` so the UI can show at-a-glance stats
 * without fetching every record.
 */
export interface ChildSummary {
  /** Unique identifier of the child. */
  childId: string;

  /** Display name. */
  name: string;

  /** Date of birth in YYYY-MM-DD format. */
  birthDate: string;

  /** Biological sex. */
  gender: 'male' | 'female';

  /** Total number of growth records for this child. Used to display "12 records" badges in the UI. */
  recordCount: number;

  /** Date of the most recent growth record, or `undefined` if no records exist yet. Used for "last updated N days ago" display. */
  lastRecordDate?: string;
}

File: python-server/app/models/growth.py
"""
成长记录模块的数据模型（Pydantic）

## 为什么字段名用驼峰（birthDate / childId）而不是 Python 惯例的蛇形？

这些模型直接对齐前端 `client/src/types` 里的 TypeScript 契约（从原 Express
后端的 Zod schema 迁移而来）。前端 `api.ts` 已经按驼峰字段发送/读取，例如：
  - 请求体：`{ name, birthDate, gender }`
  - 响应体：`{ childId, recordCount, lastRecordDate }`

保持驼峰字段名可以让前端零改动接入，代价是后端代码不严格遵循 PEP8 的蛇形
命名。这是「契约一致性优先」的务实取舍。

## 校验边界（对照原 Zod schema）

WHO 0-5 岁儿童生长标准范围：
  - 身高 height：30-180 cm
  - 体重 weight：1-100 kg
  - 头围 headCircumference：20-80 cm
  - 睡眠 sleepDuration：0-24 h
  - 尿布 diapers：0-50 次/天
"""

from typing import Literal, Optional

from pydantic import BaseModel, Field


class Feeding(BaseModel):
    """单次喂养记录"""
    type: Literal["breast", "formula", "solid", "mixed"]
    amount: Optional[float] = None
    unit: Optional[Literal["ml", "oz", "g", "次"]] = None
    notes: Optional[str] = Field(None, max_length=200)


class CreateChild(BaseModel):
    """新建宝宝档案的请求体"""
    name: str = Field(..., min_length=1, max_length=50)
    birthDate: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    gender: Literal["male", "female"]


class UpdateChild(BaseModel):
    """更新宝宝档案（部分更新，字段全可选）"""
    name: Optional[str] = Field(None, min_length=1, max_length=50)
    birthDate: Optional[str] = Field(None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    gender: Optional[Literal["male", "female"]] = None


class CreateGrowthRecord(BaseModel):
    """新增成长记录的请求体（除 date 外均可选）"""
    date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    height: Optional[float] = Field(None, ge=30, le=180)
    weight: Optional[float] = Field(None, ge=1, le=100)
    headCircumference: Optional[float] = Field(None, ge=20, le=80)
    sleepDuration: Optional[float] = Field(None, ge=0, le=24)
    feeding: Optional[Feeding] = None
    diapers: Optional[int] = Field(None, ge=0, le=50)
    notes: Optional[str] = Field(None, max_length=500)


class UpdateGrowthRecord(BaseModel):
    """更新成长记录（部分更新，字段全可选）"""
    date: Optional[str] = Field(None, pattern=r"^\d{4}-\d{2}-\d{2}$")
    height: Optional[float] = Field(None, ge=30, le=180)
    weight: Optional[float] = Field(None, ge=1, le=100)
    headCircumference: Optional[float] = Field(None, ge=20, le=80)
    sleepDuration: Optional[float] = Field(None, ge=0, le=24)
    feeding: Optional[Feeding] = None
    diapers: Optional[int] = Field(None, ge=0, le=50)
    notes: Optional[str] = Field(None, max_length=500)


class GrowthRecord(CreateGrowthRecord):
    """持久化后的成长记录（含服务端生成的 id + createdAt）"""
    id: str
    createdAt: str


class Child(BaseModel):
    """宝宝完整档案（含全部成长记录）"""
    childId: str
    name: str
    birthDate: str
    gender: Literal["male", "female"]
    records: list[GrowthRecord] = []
    createdAt: str
    updatedAt: str


class ChildSummary(BaseModel):
    """列表页用的轻量投影（不含 records，带统计字段）"""
    childId: str
    name: str
    birthDate: str
    gender: Literal["male", "female"]
    recordCount: int
    lastRecordDate: Optional[str] = None
